library/std/src/io/buffered/bufwriter.rs

```diff
@@ -150,7 +155,11 @@ impl<W: Write> BufWriter<W> {
         impl Drop for BufGuard<'_> {
             fn drop(&mut self) {
                 if self.written > 0 {
-                    self.buffer.drain(..self.written);
+                    if self.done() {
+                        self.buffer.clear();
+                    } else {
+                        self.buffer.drain(..self.written);
```

Comment:
Why not `truncate`? from description in https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.truncate it looks like `truncate` should do it better, but better need test it.

Reply:
Ah, there we drop elements from left, but truncate drops them from right.